### ISSUE TITLE
JPA 2.x unidirectional OneToMany relationship database

### ISSUE THREAD

**Issue by rsoika** [imixs/imixs-workflow#191]
If we use a @JoinColumn for OneToMany statements we can change and simplify  the database schema.

https://en.wikibooks.org/wiki/Java_Persistence/OneToMany

Entity Class - change getter:

```
	@OneToMany(fetch = FetchType.LAZY)
	@JoinColumn(name="ENTITY_ID", referencedColumnName="ID")
	public List<TextItem> getTextItems() {
		if (textItems == null)
			textItems = new Vector<TextItem>();
		return textItems;
	}

	public void setTextItems(List<TextItem> textItems) {
		this.textItems = textItems;
	}
```

TextItem Class - add attribute:

```
	@Column(name="ENTITY_ID")
	private String entityId;
```

As a result the following JPQL statement:

```
SELECT entity FROM Entity entity 
JOIN entity.textItems t1
JOIN entity.textItems t2
WHERE entity.type='workitem' 
AND t1.itemName="$modelversion" AND t1.itemValue="1.0.1"
AND t2.itemName="namcreator" AND t2.itemValue="admin"
ORDER BY entity.modified DESC
```

can be replaced without JOINs but sub-selects:

```
SELECT entity FROM Entity entity 
WHERE entity.type='workitem' 
  AND entity.id IN (
	SELECT t.entityId FROM TextItem AS t
 	WHERE t.itemName="$modelversion" AND t.itemValue="1.0.1"
      )
  AND entity.id IN (
	SELECT t.entityId FROM TextItem AS t
 	WHERE t.itemName="namcreator" AND t.itemValue="admin"
      )
ORDER BY entity.modified DESC
```

**Comment by rsoika** [comment]
Analyse of PostgreSQL:

```
SELECT entity FROM Entity entity 
JOIN entity.textItems t1
JOIN entity.textItems t2
WHERE entity.type='workitem' 
AND t1.itemName="$modelversion" AND t1.itemValue="1.0.1"
AND t2.itemName="namcreator" AND t2.itemValue="admin"
ORDER BY entity.modified DESC


SQL:
SELECT DISTINCT t1.ID AS a1, t1.CREATED AS a2, t1.MODIFIED AS a3, t1.TYPE AS a4, t1.VERSION AS a5
 FROM TEXTITEM t0, TEXTITEM t2, ENTITY t1
 WHERE ((((((t1.TYPE = $1)
	 AND (t0.ITEMNAME = $2))
	 AND (t0.ITEMVALUE = $3))
	 AND (t2.ITEMNAME = $4))
	 AND (t2.ITEMVALUE = $5))
	 AND ((t0.ENTITY_ID = t1.ID)
	 AND (t2.ENTITY_ID = t1.ID)))
 ORDER BY t1.MODIFIED DESC LIMIT $6 OFFSET $7



Without Joins:
====================

SELECT entity FROM Entity entity 
WHERE entity.type='workitem' 
  AND entity.id IN (
	SELECT t.entityId FROM TextItem AS t
 	WHERE t.itemName="$modelversion" AND t.itemValue="1.0.1"
      )
  AND entity.id IN (
	SELECT t.entityId FROM TextItem AS t
 	WHERE t.itemName="namcreator" AND t.itemValue="admin"
      )
ORDER BY entity.modified DESC




SQL:

SELECT DISTINCT t0.ID AS a1, t0.CREATED AS a2, t0.MODIFIED AS a3, t0.TYPE AS a4, t0.VERSION AS a5
 FROM ENTITY t0
 WHERE (((t0.TYPE = $1) 
 AND t0.ID IN (
	SELECT t1.ENTITY_ID FROM TEXTITEM t1 WHERE ((t1.ITEMNAME = $2) AND (t1.ITEMVALUE = $3))))
 AND t0.ID IN (
	SELECT t2.ENTITY_ID FROM TEXTITEM t2 WHERE ((t2.ITEMNAME = $4) AND (t2.ITEMVALUE = $5))))
 ORDER BY t0.MODIFIED DESC LIMIT $6 OFFSET $7
```

**Comment by rsoika** [comment]
As a result of further testing, we come to the conclusion, that we can not optimize much on the database side by changing the OneToMany relationships or fetch types. Performance slows down in case of a lot of equal key-value pairs in the table textItem (e.g. more than 10.000 entries of same $modelVersion).
But after all we see that Lucene is still able to solve such complex queries very fast. 

New approach for version 4.0.0:
--------------------------
We will follow now a complete new approach for version 4.0.0:
 * Lucene becomes an integral part of the imixs-workflow engine (EntityService). 
 * We will drop all oneToMany relationships from the Entity JPA Bean.
 * The EntityService will index each entity during the save method (lucene)
 * We can skip the lucene -adapter project and also the lucen plugin
 * We provide new Finder Methods in the EntiyService using the Lucene Query Builder
 * We provide a migration service bean which implodes entities finally and adds them into the lucene index.

Finally we need to refactor all JPQL statements used until now. 
Also we need an admin interface to re-index entities on some simple criterias (e.g. type, $modified, $created)

**Comment by rsoika** [comment]
Closed - see issue #192
